data/changelog/posts/ocaml-lsp/2025-08-20-ocaml-eglot-brings-lsp-support-to-emacs.md

```diff
@@ -0,0 +1,143 @@
+
+**Use ocaml-eglot if you:**
+- Are starting fresh with OCaml in Emacs
+- Want simplified configuration with automatic setup
+- Use multiple editors and want consistent OCaml support
+- Want access to project-wide search and rename features
+
+**Stick with merlin.el if:**
+- Your current setup is working perfectly and heavily customized
+- You prefer direct Merlin communication
+
+Both approaches are fully supported.
+
+## Key Benefits
+
+- **Simplified setup**: Install package, add config, start coding
+- **Same features**: All `merlin.el` functionality with identical keybindings
+- **Enhanced capabilities**: Project-wide rename, type-based search, automatic formatting
+- **Platform integration**: Works seamlessly with opam, dune, ocamlformat, and other OCaml tools
+
+## Getting Started
+
+### Prerequisites
+- Emacs 29.1+ (includes Eglot)
+- `opam install ocaml-lsp-server ocamlformat`
+
+### Basic Setup
+
+```elisp
+(use-package tuareg
+  :ensure t
+  :mode (("\\.ocamlinit\\'" . tuareg-mode)))
+
+(use-package ocaml-eglot
+  :ensure t
+  :after tuareg
+  :hook
+  (tuareg-mode . ocaml-eglot)
+  (ocaml-eglot . eglot-ensure))
+```
+
+### Recommended Configuration
+
+```elisp
+;; Better error display
+(use-package flymake
+  :ensure t
+  :config
+  (setq flymake-diagnostic-format-alist
+        '((t . (origin code message)))))
+
+;; OCaml support with formatting
+(use-package tuareg :ensure t)
+(use-package dune :ensure t)
+(use-package opam-switch-mode
+  :ensure t
+  :hook (tuareg-mode . opam-switch-mode))
+
+(use-package ocaml-eglot
+  :ensure t
+  :after tuareg
+  :hook
+  (tuareg-mode . ocaml-eglot)
+  (ocaml-eglot . eglot-ensure)
+  (ocaml-eglot . (lambda () 
+                   (add-hook #'before-save-hook #'eglot-format nil t))))
+```
+
+## Essential Commands
+
+| Feature | Command | Key Binding |
+|---------|---------|-------------|
+| Show type | `ocaml-eglot-type-enclosing` | `C-c C-t` |
+| Jump to definition | `ocaml-eglot-find-definition` | `C-c C-l` |
+| Jump to declaration | `ocaml-eglot-find-declaration` | `C-c C-i` |
+| Generate patterns | `ocaml-eglot-destruct` | `C-c \|` |
+| Fill holes | `ocaml-eglot-construct` | `C-c \` |
+| Switch .ml/.mli | `ocaml-eglot-alternate-file` | `C-c C-a` |
+| Show documentation | `ocaml-eglot-document` | `C-c C-d` |
+| Search by type | `ocaml-eglot-search` | |
+| Rename symbol | `ocaml-eglot-rename` | |
+
+## Migration from merlin.el
+
+1. Install: `opam install ocaml-lsp-server ocamlformat`
+2. Replace your Merlin configuration with the `ocaml-eglot` setup above
+3. Restart Emacs
+
+Your existing keybindings work immediately.
+
+## Enhanced Features Examples
+
+**Type-based search**: Find functions by signature
+```
+Search for: string -> int option
+Result: String.to_int, Int.of_string, etc.
+```
+
+**Project-wide rename** (OCaml 5.2+):
+```bash
+# Build index first
+dune build @ocaml-index
+# Then use ocaml-eglot-rename in Emacs
+```
+
+**Polarity search**: Find functions by input/output types
```

Comment:
The function is also available in `merlin.el`

Reply:
(And I think that search by types is better, search by polarity is historically present).